allennlp/data/iterators/data_iterator.py

```diff
@@ -53,26 +54,26 @@ def __call__(self,
             for _ in range(num_epochs):
                 yield from self._yield_one_epoch(dataset, shuffle, cuda_device, for_training)
 
-    def get_num_batches(self, dataset: Dataset) -> int:
+    def get_num_batches(self, dataset: DatasetType) -> int:
         """
         Returns the number of batches that ``dataset`` will be split into; if you want to track
         progress through the batch with the generator produced by ``__call__``, this could be
         useful.
         """
         raise NotImplementedError
 
-    def _yield_one_epoch(self, dataset: Dataset, shuffle: bool, cuda_device: int, for_training: bool):
+    def _yield_one_epoch(self, dataset: DatasetType, shuffle: bool, cuda_device: int, for_training: bool):
         grouped_instances = self._create_batches(dataset, shuffle)
         for group in grouped_instances:
-            batch = Dataset(group)
+            batch = InMemoryDataset(group)
             padding_lengths = batch.get_padding_lengths()
             logger.debug("Batch padding lengths: %s", str(padding_lengths))
             logger.debug("Batch size: %d", len(batch.instances))
             yield batch.as_tensor_dict(padding_lengths,
                                        cuda_device=cuda_device,
                                        for_training=for_training)
 
-    def _create_batches(self, dataset: Dataset, shuffle: bool) -> List[List[Instance]]:
+    def _create_batches(self, dataset: DatasetType, shuffle: bool) -> Iterator[List[Instance]]:
```

Comment:
Maybe have this return an `Iterator[InMemoryDataset]`?  It means you can't be as fancy with the `islice` stuff you were doing, but I _think_ that doesn't work anyway, if you want to allow for a `LazyBucketIterator`.  Making `Iterator[InMemoryDataset]` the API encourages the right design decision when implementing laziness (I think), and it makes it clear that `InMemoryDataset` is a "batch".